docs: diataxis package docs (#504)
canonical/charmlibs: pathops/docs/explanation/design.rst

.. meta::
   :description: Why pathops models charm filesystem operations on pathlib.

> Design Philosophy

Reviewer: ```suggestion
Design philosophy
```

=================

> :mod:`pathops` provides a :mod:`pathlib`-like interface for working with files in both local and remote (Pebble-based) filesystems. This document explains the design choices behind the library.

Reviewer: I think it's only necessary to link to `pathops` and `pathlib` the first time they're mentioned on the page. Removing the later links should help with readability and make the remaining links stand out better.

A pathlib-like interface
------------------------

The core goal of :mod:`pathops` is to let charm authors write filesystem code once and have it work on both Kubernetes and machine charms. To achieve this, :class:`~pathops.ContainerPath` mirrors the :class:`pathlib.PurePosixPath` and :class:`pathlib.Path` APIs as closely as possible.

> Python developers understand :mod:`pathlib`. By matching its interface — the ``/`` operator for joining, :meth:`~pathlib.Path.read_text` and :meth:`~pathlib.Path.write_text` for file I/O, :meth:`~pathlib.Path.mkdir` for directory creation, :meth:`~pathlib.Path.glob` for matching — :mod:`pathops` avoids introducing new concepts. Code that works with :class:`pathlib.Path` should look almost identical when ported to :class:`~pathops.ContainerPath`.

Reviewer: I might adjust this for easier reading:

```suggestion
``pathlib`` has a familiar interface — the ``/`` operator for joining, :meth:`~pathlib.Path.read_text` and :meth:`~pathlib.Path.write_text` for file I/O, :meth:`~pathlib.Path.mkdir` for directory creation, :meth:`~pathlib.Path.glob` for matching. By matching ``pathlib``'s interface, ``pathops`` avoids introducing new concepts. Code that works with ``pathlib.Path`` should look almost identical when ported to ``ContainerPath``.
```

- **Absolute paths only.** :class:`~pathops.ContainerPath` raises :class:`~pathops.RelativePathError` if you try to create a relative path, because Pebble operations require absolute paths.
- **No** ``chmod()`` **method.** Permissions are set at write time via ``mode=``, ``user=``, and ``group=`` keyword arguments on ``write_text()``, ``write_bytes()``, and ``mkdir()``. This matches Pebble's API, where permissions are part of the push/make_dir call.
- **Default file mode is 0o644**, not :mod:`pathlib`'s 0o666. This matches Pebble's default and is a more sensible default for most use cases.

PathProtocol
------------

:class:`~pathops.PathProtocol` is a :class:`typing.Protocol` that defines the interface common to both :class:`~pathops.ContainerPath` and :class:`~pathops.LocalPath`. It exists so that charm authors can write helpers that accept either path type:

.. code-block:: python

   from charmlibs import pathops

   def write_config(root: pathops.PathProtocol, content: str) -> None:
       (root / 'etc' / 'myapp' / 'config.yaml').write_text(content)

> You might think ``LocalPath | ContainerPath`` would work just as well, but in practice it's much less useful. A union type shows the *superset* of methods and arguments from both types in your IDE — including things like :class:`pathlib.PosixPath` methods that don't exist on :class:`~pathops.ContainerPath`. :class:`~pathops.PathProtocol` shows only the *subset* that both types support, so your IDE completions are exactly the methods you can safely call.

Reviewer: Thanks for this explanation. I didn't appreciate it until now!

.. note::

>    :class:`~pathops.PathProtocol` is not designed to be implemented by third parties. We don't guarantee backwards compatibility for new methods added to the protocol — adding a method is a minor version bump, not a major one. Use it as a type annotation, not as a base class.

Reviewer: I know what you mean, but "backwards compatibility for new methods" could be slightly confusing: if a method is new, how is there anything for it to be backwards compatible with?

How about this:

```suggestion
   :class:`~pathops.PathProtocol` is not designed to be implemented by third parties. We don't guarantee compatibility for third-party implementations when we add new methods to the protocol — adding a method is a minor version bump, not a major one. Use the protocol as a type annotation, not as a base class.
```

Author: ```suggestion
   :class:`~pathops.PathProtocol` is not designed to be implemented by third parties. We don't guarantee compatibility for third-party implementations when we add new methods to the protocol. One example: adding a method to the protocol only a minor version bump (it's a feature for users of the concrete classes), not a major one (it's a potentially breaking change for implementations of the protocol). Treat the protocol as a convenient type annotation for the ``pathops`` classes, not an compatibility guarantee for other implementations.
```

Author: Thanks, reworded further based on your suggestions.

.. note::

>    :class:`~pathops.PathProtocol` is not designed to be implemented by third parties. We don't guarantee compatibility for third-party implementations when we add new methods to the protocol. One example: adding a method to the protocol only a minor version bump (it's a feature for users of the concrete classes), not a major one (it's a potentially breaking change for implementations of the protocol). Treat the protocol as a convenient type annotation for the ``pathops`` classes, not an compatibility guarantee for other implementations.

Reviewer: ```suggestion
   :class:`~pathops.PathProtocol` is not designed to be implemented by third parties. We don't guarantee compatibility for third-party implementations when we add new methods to the protocol. One example: adding a method to the protocol is only a minor version bump (it's a feature for users of the concrete classes), not a major one (it's a potentially breaking change for implementations of the protocol). Treat the protocol as a convenient type annotation for the ``pathops`` classes, not a compatibility guarantee for other implementations.
```
